File: php/test/EtaEntityTest.php
<?php
declare(strict_types=1);

// Eta entity test

require_once __DIR__ . '/../realtimebusdata_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EtaEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = RealTimeBusDataSDK::test(null, null);
        $ent = $testsdk->Eta(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = eta_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "eta." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set REALTIMEBUSDATA_TEST_ETA_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $eta_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.eta")));
        $eta_ref01_data = null;
        if (count($eta_ref01_data_raw) > 0) {
            $eta_ref01_data = Helpers::to_map($eta_ref01_data_raw[0][1]);
        }

        // LIST
        $eta_ref01_ent = $client->Eta(null);
        $eta_ref01_match = [
            "route" => $setup["idmap"]["route01"],
            "service_type" => $setup["idmap"]["service_type01"],
        ];

        [$eta_ref01_list_result, $err] = $eta_ref01_ent->list($eta_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($eta_ref01_list_result);

        // LOAD
        $eta_ref01_match_dt0 = [];
        [$eta_ref01_data_dt0_loaded, $err] = $eta_ref01_ent->load($eta_ref01_match_dt0, null);
        $this->assertNull($err);
        $this->assertNotNull($eta_ref01_data_dt0_loaded);

    }
}

function eta_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/eta/EtaTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = RealTimeBusDataSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["eta01", "eta02", "eta03", "route_eta01", "route_eta02", "route_eta03", "stop_eta01", "stop_eta02", "stop_eta03", "route01", "service_type01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("REALTIMEBUSDATA_TEST_ETA_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "REALTIMEBUSDATA_TEST_ETA_ENTID" => $idmap,
        "REALTIMEBUSDATA_TEST_LIVE" => "FALSE",
        "REALTIMEBUSDATA_TEST_EXPLAIN" => "FALSE",
        "REALTIMEBUSDATA_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["REALTIMEBUSDATA_TEST_ETA_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["REALTIMEBUSDATA_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["REALTIMEBUSDATA_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new RealTimeBusDataSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["REALTIMEBUSDATA_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["REALTIMEBUSDATA_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/eta_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/real-time-bus-data-sdk/go"
	"github.com/voxgig-sdk/real-time-bus-data-sdk/go/core"

	vs "github.com/voxgig-sdk/real-time-bus-data-sdk/go/utility/struct"
)

func TestEtaEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Eta(nil)
		if ent == nil {
			t.Fatal("expected non-nil EtaEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := etaBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "eta." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set REAL_TIME_BUS_DATA_TEST_ETA_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		etaRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.eta", setup.data)))
		var etaRef01Data map[string]any
		if len(etaRef01DataRaw) > 0 {
			etaRef01Data = core.ToMapAny(etaRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = etaRef01Data

		// LOAD
		etaRef01Ent := client.Eta(nil)
		etaRef01MatchDt0 := map[string]any{}
		etaRef01DataDt0Loaded, err := etaRef01Ent.Load(etaRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if etaRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func etaBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "eta", "EtaTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read eta test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse eta test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"eta01", "eta02", "eta03", "route_eta01", "route_eta02", "route_eta03", "stop_eta01", "stop_eta02", "stop_eta03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("REAL_TIME_BUS_DATA_TEST_ETA_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"REAL_TIME_BUS_DATA_TEST_ETA_ENTID": idmap,
		"REAL_TIME_BUS_DATA_TEST_LIVE":      "FALSE",
		"REAL_TIME_BUS_DATA_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["REAL_TIME_BUS_DATA_TEST_ETA_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["REAL_TIME_BUS_DATA_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewRealTimeBusDataSDK(core.ToMapAny(mergedOpts))
	}

	live := env["REAL_TIME_BUS_DATA_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["REAL_TIME_BUS_DATA_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
